src/utils.rs

```diff
@@ -0,0 +1,40 @@
+use bincode::{self, Options};
+use serde::{Deserialize, Serialize};
+use std::fmt::Write;
+
+use crate::error::Result;
+
+pub fn serialize<T>(object: &T) -> Result<Vec<u8>>
+where
+    T: Serialize,
+{
+    let opts = bincode::DefaultOptions::new()
+        .with_little_endian()
+        .with_fixint_encoding();
+    Ok(opts.serialize(&object)?)
+}
+
+pub fn deserialize<'a, T>(buffer: &'a Vec<u8>) -> Result<T>
+where
+    T: Deserialize<'a>,
+{
+    let opts = bincode::DefaultOptions::new()
+        .with_little_endian()
+        .with_fixint_encoding();
+    let decoded: T = opts.deserialize(buffer.as_slice())?;
+
+    Ok(decoded)
+}
+
+pub fn arr_to_hex(bytes: &[u8]) -> String {
+    bytes.iter().fold(String::new(), |mut output, b| {
+        let _ = write!(output, "{b:02X}");
+        output
+    })
+}
+
+pub fn arr_to_string(arr: &[u8]) -> String {
+    let len = arr.iter().position(|&x| x == 0).unwrap_or(arr.len());
+    let valid_bytes = &arr[..len];
+    String::from_utf8(valid_bytes.to_vec()).expect("utf8 conversion error")
```

Comment:
Do you intend to convert a c-string (null terminated array) to a rust string? If so take a look at this https://doc.rust-lang.org/std/ffi/struct.CString.html

Reply:
The string might be not null-terminated (the case where the whole slice is filled up with characters). Also, the CString require unsafe code blocks, so I would prefer to leave this function as it is.